### PR TITLE
Add API for emitting compact unwind encoding

### PR DESCRIPTION
**Pull request by filipnavara**
…enforce DWARF encoding if not explicitly overridden

Currently `objwriter` will do best effort to convert DWARF CFI codes into compact unwinding encoding. On osx-arm64 we always fail this due to the way NativeAOT code generator emits prologs. It falls back into the "use DWARF CFI encoding" which the linker turns into efficient index. On osx-x64 we never hit the code since ILCompiler doesn't specify macOS/Darwin version in the triple and thus compact unwinding is not generated by objwriter. The conversion is then done by the ld64 linker where we have only limited control over what is produced.

With the change we get new options:
- By default `objwriter` will always use the "use DWARF CFI encoding" value in the compact unwinding tables. That ensures we have DWARF CFI info available and that we can unwind function prologs and epilogs.
- New `EmitCFICompactUnwindEncoding` API is introduced which can force specific encoding.
  - If the parameter is set to `0` it will switch to the previous behavior and let LLVM try to convert DWARF CFI into compact unwinding.
  - If the parameter is set to anything but zero it's assumed to be the architecture specific compact unwinding encoding. The code generator can produce it more efficiently and reliably than the conversion from DWARF CFI.

I don't currently intend to use the `EmitCFICompactUnwindEncoding` API on NativeAOT side but it may change in future if we decide to produce compatible prologs/epilogs and use compact unwinding. On osx-x64 the compact unwinding is still not generated by default unless `darwin10` (or newer) is used in the triple. I intend NativeAOT to do that.

### CODE REVIEW

**Comment by filipnavara** [comment]
cc @MichalStrehovsky

**Comment by MichalStrehovsky** [comment]
> cc @MichalStrehovsky

I'll wait for @markples to come back to review this. I know nothing about unwinding. ObjWriter ownership is getting transitioned to the JIT team since the JIT team knows about unwinding :).

**Comment by filipnavara** [comment]
> > cc @MichalStrehovsky
> 
> I'll wait for @markples to come back to review this. I know nothing about unwinding. ObjWriter ownership is getting transitioned to the JIT team since the JIT team knows about unwinding :).

As long as someone will look at it I am fine, thanks for update! I will also be happy to answer any questions about it.

The gist is that there are two possible unwinding encodings used in Mach-O - DWARF CFI (also used by ELF on Linux) and compact unwinding. Compact unwinding is a concise format that describes the common frame format (how original SP and other registers were saved to stack in prolog) using single 32-bit integer per method (or funclet in case of NativeAOT). This is very efficient in terms of size but it's primarily meant for unwinding synchronous exceptions that happen in method bodies. The format lacks description of how to unwind method prologs/epilogs which is useful during asynchronous exceptions. Specifically for NativeAOT it's useful when method is interrupted by signal to have its return address hijacked, which can happen in prolog or epilog.

Secondary function of the compact unwinding format is that it acts as efficient lookup table for translation between address and unwinding information. The DWARF CFI format itself lacks any kind of index. ELF files on Linux use an extra `.eh_frame_hdr` section with an index. The Apple ld64 linker knows only the compact unwinding index though.

.NET JIT emits the DWARF CFI format. We want to preserve it to keep the ability to unwind methods at any execution point. However, we also want to generate the compact unwinding table as the index for performance reasons. This PR gives fine-grained control to allow that.

Short-term this is about ensuring correct and precise output with index that guarantees usable performance at runtime (on par with Linux/ELF).

Long-term we may want to find a way to use the concise unwinding information to save on executable size. This would either mean limiting thread hijacking, optionally using conservative GC mode, or manually unwinding the prolog by interpreting effects of known instruction in the code (already done for epilog in NativeAOT).

**Comment by jkotas** [comment]
cc @VSadov

File: llvm/tools/objwriter/objwriter.h
Comment (VSadov) [inline llvm/tools/objwriter/objwriter.h:103]: So, as I understand - the idea is to switch the default to be DWARF, and this will be a way to request compact format, if so desired in the future.

Comment (filipnavara) [reply llvm/tools/objwriter/objwriter.h:103]: Correct

File: llvm/tools/objwriter/objwriter.cpp
Comment (VSadov) [inline llvm/tools/objwriter/objwriter.cpp:599]: I think I need a bit more explanation on what `CompactUnwindDwarfEHFrameOnly` means.

The comment says:
```c
/// Compact unwind encoding indicating that we should emit only an EH frame.
```

, but how is it significant here?

**Comment by markples** [comment]
> The gist is...

Thank you @filipnavara for the explanation.  I've just started looking but wanted to at least acknowledge the PR.

File: llvm/tools/objwriter/objwriter.h
Comment (markples) [reply llvm/tools/objwriter/objwriter.h:103]: Could you please document this in the code?

(However, it is not clear to me yet if I'm asking for a quick comment here about the behavior of (zero -> attempt, non-zero -> provided) or more of your general explanation to be included somewhere else.  I'll continue reading but am certainly open to a suggestion.)

File: llvm/tools/objwriter/objwriter.cpp
Comment (filipnavara) [reply llvm/tools/objwriter/objwriter.cpp:599]: It means the compact unwind value is not available and stores the offset into the DWARF CFI (data in the `EH frame` section) instead. In effect, it works as index only.

File: llvm/tools/objwriter/objwriter.h
Comment (filipnavara) [reply llvm/tools/objwriter/objwriter.h:103]: Sure. I will try to do that tomorrow.

**Comment by markples** [comment]
cc @BrianBohe

**Review by VSadov — APPROVED** [review]
LGTM. Thanks!

File: llvm/include/llvm/MC/MCStreamer.h
Comment (markples) [inline llvm/include/llvm/MC/MCStreamer.h:1006]: A lot of these emitCFI* virtuals are overridden in MCAsmStreamer.cpp.  Does that make sense here?  I'm not familiar enough with these yet, so I think I'm better off asking you than waiting until I can learn it all.

File: llvm/tools/objwriter/objwriter.h
Comment (markples) [reply llvm/tools/objwriter/objwriter.h:103]: Thanks!  To verify my understanding:

The Streamer is changed from "always try to infer" to "infer if zero".

The objwriter abstraction on top of it changes to a default of the `...EHFrameOnly` value (which is not zero as defined in the header file but the specific `UNWIND_...` value written in `MCObjectFileInfo.cpp`).  That can be overridden to zero (old default - infer) or any non-zero value (just use it).

I assume that any producer has to be aware of the special values such as `UNWIND_...` and use a different range of numbers.

**Review by markples — APPROVED** [review]
Thank you for this change.  LGTM.  We can merge this to (finally, sorry this took so long on my side) unblock the other work.  If the asm stream needs an update, it can go in separately.

**Comment by VSadov** [comment]
Thanks!!

File: llvm/include/llvm/MC/MCStreamer.h
Comment (filipnavara) [reply llvm/include/llvm/MC/MCStreamer.h:1006]: The unwinding encoding is inherently specific to the Apple Mach-O file format and a processor architecture target. Any value for `Encoding` parameter besides 0 (ie. letting LLVM use the architecture specific conversation which uses the complex layering and virtual overrides) would have to be generated in the NativeAOT ILCompiler with the knowledge of specific target (ie. the values have different meaning on x64 and Arm64).

There are certain function prolog patterns that can be expressed by compact unwinding but which LLVM cannot infer automatically. For example, when there is more than one page of stack reserved space the JIT may generate a `__chkstk` call to probe the stack pages. The compact unwinding code in that case is "RBP-based frame with indirect stack size" type. This unwind code works by encoding the offset of `<value>` in the `sub rsp, <value>` instruction to determine the frame size. However, since the `__chkstk` pattern uses different instructions LLVM doesn't recognize it. The ILCompiler/JIT can recognize it and compose the value of the unwind code by itself because it knows these patterns. Obviously the code would have to be highly specific for both the osx-x64 and osx-arm64 targets. In the end it is an optimization that may or may not be worth the effort. However, since we know what the JIT emits it's much easier to implement in ILCompiler than by LLVM trying to second guess it (and guessing it wrong leads to runtime crashes that are hard to debug).

Comment (filipnavara) [reply llvm/include/llvm/MC/MCStreamer.h:1006]: Here's an example of prolog that LLVM cannot handle by itself:
```
0000000000000000 55                              push      rbp
0000000000000001 4157                            push      r15
0000000000000003 4156                            push      r14
0000000000000005 4155                            push      r13
0000000000000007 4154                            push      r12
0000000000000009 53                              push      rbx
000000000000000A 4C8D9C2468EFFFFF                lea       r11,[rsp-1098h]
0000000000000012 E800000000                      call      0000000000000017h // __chkstk
0000000000000017 498BE3                          mov       rsp,r11
```

On second inspection I see that this would need a codegen change to get positive offset in the code to generate valid compact encoding. The `<value>` in `lea       r11,[rsp-1098h]` is negative offset which cannot be encoded. It would be encodable as compact unwinding if the last few instructions were

```
mov r11, rsp
sub r11, 1098h
call __chkstk
mov rsp, r11
```

Comment (markples) [reply llvm/include/llvm/MC/MCStreamer.h:1006]: Sorry, my question was obviously very unclear (though separately it was still useful to see your explanation and example).

This diff adds a new virtual function to `MCStreamer`.  `MCAsmStreamer` is a derived class of `MCStreamer` used for writing textual assembly files.  The other `emitCFIxxx` functions have overrides in `MCAsmStreamer`, which suggests that this function also should affect the asm output.  I don't know if that means an `emitCFICompactUnwindEncoding` override that simply adds a directive or if something more complicated would need to be done.  If the correct way isn't an override, then perhaps `emitCFICompactUnwindEncoding` doesn't need to be virtual.

Comment (filipnavara) [reply llvm/include/llvm/MC/MCStreamer.h:1006]: My bad :-) I wanted to explain that but got side-tracked with my line of thinking...

You are right, the `emitCFICompactUnwindEncoding` method doesn't have to be virtual. In fact, the only reason it needs to exist is that `getCurrentDwarfFrameInfo` is private and `getDwarfFrameInfos` returns `const` reference, so neither of those existing methods can be used by objwriter directly.

**Pull request merged**
